File: 2024/day1.cpp
#include <bits/stdc++.h>
#include <fstream>

using namespace std;

pair<int, int> getNums(string s) {
    // space delimiter size: 3
    int idx = s.find(" ");
    return make_pair(stoi(s.substr(0, idx)), stoi(s.substr(idx+3)));
}

int main() {
    ifstream inputFile("./input/day1.txt");
    string nums;
    vector<int> list1, list2;
    map<int, int> freq;
    while(getline(inputFile, nums)) {
        auto [a, b] = getNums(nums);
        list1.push_back(a);
        list2.push_back(b);
        freq[b]++;
    }
    inputFile.close();
    sort(list1.begin(), list1.end());
    sort(list2.begin(), list2.end());
    int n = list1.size();   assert(list2.size() == n);

    long long sumOfDistances = 0, similarityScore = 0;
    for(int i=0; i<n; ++i) {
        sumOfDistances += abs(list1[i] - list2[i]);
        similarityScore += list1[i] * freq[list1[i]];
    }
    cout<< "par1: "<< sumOfDistances<< endl;
    cout<< "part2: "<< similarityScore<< endl;
}

File: 2024/day1.scala
import scala.collection.mutable.ArrayBuffer
import scala.io.Source.fromFile

object Main {
  def main(args: Array[String]): Unit = {
    val list1 = ArrayBuffer[Int]()
    val list2 = ArrayBuffer[Int]()

    val inputFile = fromFile("./input/day1.txt")
    try {
        val lines = inputFile.getLines
        while(lines.hasNext) {
            val Array(num1, num2) = lines.next.split("   ")
            list1 += num1.toInt
            list2 += num2.toInt
        }
    } finally inputFile.close
    
    val zippedList = list1.sorted zip list2.sorted
    val sumOfDistances = zippedList.map(x => (x(0) - x(1)).abs).sum
    println(s"part1: $sumOfDistances")

    val freq = list2.groupMapReduce(num => num)(_ => 1)(_ + _)
    val similarityScore = list1.map(num => num * freq.getOrElse(num, 0)).sum
    println(s"part2: $similarityScore")
  }
}